Remotion/Web/Development.WebTesting/WebDriver/Configuration/Firefox/FirefoxBinariesProvider.cs

```diff
@@ -0,0 +1,207 @@
+      // meaning the newest version must be found and selected.
+      var installedFirefoxSubKeys = Registry.LocalMachine.OpenSubKey (c_firefoxRegistryPath)?.GetSubKeyNames();
+      var latestFirefoxSubKeyName = installedFirefoxSubKeys?.OrderByDescending (x => new Version (x.Split (' ')[0])).FirstOrDefault();
+
+      const string errorMessage = "Installed Firefox path could not be read from the registry.";
+      Assertion.IsNotNull (latestFirefoxSubKeyName, errorMessage, c_firefoxRegistryPath);
+
+      var installedFirefoxRegistryKey = Registry.LocalMachine.OpenSubKey (Path.Combine (c_firefoxRegistryPath, latestFirefoxSubKeyName, "Main"));
+      Assertion.IsNotNull (installedFirefoxRegistryKey, errorMessage, c_firefoxRegistryPath);
+
+      return installedFirefoxRegistryKey.GetValue ("PathToExe").ToString();
+    }
+
+    private string GetDriverPathAndDownloadIfMissing (string firefoxPath)
+    {
+      var firefoxVersion = GetFileVersion (firefoxPath);
+
+      if (firefoxVersion < s_minimumSupportedFirefoxVersion)
+      {
+        throw new NotSupportedException (
+            string.Format (
+                "The installed Firefox version ({0}) is lower than the minimum required version of {1}.",
+                firefoxVersion.ToString (1),
+                s_minimumSupportedFirefoxVersion.ToString (1)));
+      }
+
+      var driverReleaseInfo = GetLatestGeckoDriverReleaseInfo();
+
+      var driverDownloadUrl = GetBrowserDownloadUrl (driverReleaseInfo);
+      var driverVersion = GetDriverVersion (driverReleaseInfo);
+
+      var versionedGeckoDriverDirectory = GetVersionedDriverDirectory (driverVersion);
+
+      if (!DriverExists (driverVersion))
+        DownloadDriver (driverDownloadUrl, versionedGeckoDriverDirectory);
+
+      return GetDriverPath (driverVersion);
+    }
+
+    private string GetDriverVersion (GithubResponse driverReleaseInfo)
+    {
+      var driverVersion = driverReleaseInfo.TagName?.TrimStart ('v');
+      return Assertion.IsNotNull (driverVersion, "Could not fetch the driver's version from GitHub. This could mean GitHub's API has changed.");
+    }
+
+    private string GetBrowserDownloadUrl (GithubResponse driverReleaseInfo)
+    {
+      var driverDownloadUrl = driverReleaseInfo.Assets
+          .Select (asset => asset.BrowserDownloadUrl)
+          .Where (url => url != null)
+          .SingleOrDefault (url => url.Contains ($"win{c_driverBitness}"));
+
+      return Assertion.IsNotNull (
+          driverDownloadUrl,
+          string.Format (
+              "Could not fetch the driver's download URL from GitHub. This could mean GitHub's API has changed, or the release does not contain the win{0} archive.",
+              c_driverBitness));
+    }
+
+    private GithubResponse GetLatestGeckoDriverReleaseInfo ()
+    {
+      using (var webClient = new WebClient())
+      {
+        webClient.Headers.Add ("user-agent", "re-motion");
```

Comment:
we really need a user agent? Is that also there with the other providers? can we remove that in all providers (not in this branch)?

Reply:
The user agent is required for github requests. Other providers do not need the agent set as they do not issue github requests.